packages/eslint-plugin/src/rules/unbound-method.ts

```diff
@@ -152,69 +147,162 @@ export default createRule<Options, MessageIds>({
               : 'unbound',
           node,
         });
+        return true;
       }
+      return false;
     }
 
-    return {
-      MemberExpression(node: TSESTree.MemberExpression): void {
-        if (isSafeUse(node)) {
-          return;
-        }
-
-        const objectSymbol = services.getSymbolAtLocation(node.object);
+    function isNativelyBound(
+      object: TSESTree.Node,
+      property: TSESTree.Node,
+    ): boolean {
+      if (
+        object.type === AST_NODE_TYPES.Identifier &&
+        property.type === AST_NODE_TYPES.Identifier
+      ) {
+        const objectSymbol = services.getSymbolAtLocation(object);
+        const notImported =
+          objectSymbol && isNotImported(objectSymbol, currentSourceFile);
 
         if (
-          objectSymbol &&
-          nativelyBoundMembers.has(getMemberFullName(node)) &&
-          isNotImported(objectSymbol, currentSourceFile)
+          notImported &&
+          nativelyBoundMembers.has(`${object.name}.${property.name}`)
```

Comment:
```ts
const { log } = console;
```
In our current configuration, the `console.log` declarations are under `@types/node`. So if we remove this check, the rule will report on this case. Not sure what's the best solution there..

Reply:
> you'd like to get rid of the first half of this function and only have the part after the `return` statement, but that that would cause bugs to do so?

Yup, exactly!

I'd prefer to use only type checking here for obvious reasons, as it is more flexible and robust than just checking identifier names.

But this type checking-based approach works by checking whether a method is defined in the default library. This can be a problem for us, especially in the case of `console` methods.

Imagine the following situation:

```jsonc
// tsconfig.json
{
  "compilerOptions": {
    "lib": ["ESNext"],
    "types": ["node"]
  }
}
```

```ts
// index.ts
const { log } = console
```

The `log` method of the `Console` interface is defined in the `node_modules/@types/node/console.d.ts`, not in the default TypeScript library. Therefore, the rule will false-positively report on this destructuring because it knows that this method is defined outside the default lib.

Reply:
Hmm. I'm not sure I have enough knowledge to know what to do here. Maybe someone from @typescript-eslint/triage-team will have a suggestion?

Reply:
Just to make sure I understand: we're saying that:

1. `console`'s `log` is a known exception defined in `lib.dom.d.ts`
2. A `log` override is defined in `@types/node`'
3. We'd like to still allow the exception for `console`'s `log`, even though type info thinks it's a `@types/node` thing

...is that right?

Reply:
If it's intentional that you can `const { log } = console; log("...");`, then `log` being defined as a _method_ rather than a _property_ is a typings bug. Both in browsers and Node. We can verify that the unbound call is allowed in both - so both have a bug.

The `"DOM"` types have exceptions for built-in bugs like this because:

* The TypeScript team is unwilling to fix many of the dom type issues because that risks unexpected changes
    * I can't find a reference for one we've reported quickly, but #18286 is a general example
* They're built-in -> so we know they're tied to TypeScript releases

But, `@types/` packages from DefinitelyTyped I'm under the impression we've traditionally leaned towards modifying:

* It's much easier to change them ad-hoc: e.g. https://github.com/DefinitelyTyped/DefinitelyTyped/issues/24193
* They're a bit closer to a third-party extension in `node_modules/` (albeit one with built-in handling) than true built-in types

The fact that [`@typescript-eslint/unbound-method` doesn't report on `@types/node` with `lib: ["esnext"]`(https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=JYWwDg9gTgLgBAbzgGwgczgXzgMyhEOAIgDsIATAUwC4BjCEgZwmUqIG4AoT1NACgCU7IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkqyARhWANqTolHoAeekAugBpwNSCTyZk0PADkArqlQBhABbpoAazKUJc9CIC%2BIA0A&tokens=false) I would even say is a _bug_ we should follow up on:

```ts
import { log } from "node:console";

log();
```

Thoughts?

Reply:
Yeah, that makes sense. So, I guess the plan would be as follows:

- Make exceptions only for known methods from the default TS library.
- Remove purely syntactic checks so that we only rely on type-level checks
- For now, treat types coming from `@types/*` as invalid and let the rule report them
- Report invalid types on the DT side

Am i right?

---

I wonder if this would be too breaking a change for people, though. We'll make them update `@types/node` to get rid of lint errors in Node.js projects.

Reply:
I would say that _Report invalid types on the DT side_ should be done ASAP, so folks keep stay up-to-date-ish dev dependencies are less likely to be impacted. And maybe delaying merging the change on our end till a couple of months after that? Would that be overkill?

Reply:
My opinion - keep the first check so we can merge without regression (even though it's "not ideal"). File DT issue, and put a link to it in code with an explanation that the first check is necessary due that issue.

Reply:
Oops, I just realized that `Console` is an interface, and we can't rely on function signatures declared in interfaces, because classes implementing those interfaces can still access `this` even if `this: void` is declared in an interface:

```ts
interface Aaa {
  unbound: (this: void) => void
}

class Foo implements Aaa {
  data = 'string'
  unbound() {
    console.log(this.data) // runtime error!
  }
}

// not reported by unbound-method rule
const { unbound } = {} as Aaa
unbound()
```

[playground](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAgIJzsg3gKGcgVxACMB7IgEwC5kAKMAC2AGcaA3U4CgSmQF4AfMg5ccAXxw4EAGzjNmyAGKlSyYAFsADtIjqI4Beky58FOGEx9kAcmZgooAObW8hEuRAVavE-mQJSEGZSHQA6aVJHeiZmULMLXgB6ROQoIjANFGgoUigAQlcJCRxk5BBSMFSITVzICmRiAE83MkoAWj1GUnq0nSlAu2wWj3qxfmwxuTQMHCJWz28gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=&tokens=false)

This is a bug in a rule. See https://github.com/typescript-eslint/typescript-eslint/issues/3503. The rule should rely on function signatures only when dealing with classes, not with interfaces.

So even if we ask DT to change the types on their side, we have to make exceptions for those types because they do not guarantee runtime safety.

---

So I think now we have two options from https://github.com/typescript-eslint/typescript-eslint/pull/8952#discussion_r1657523233:

> 1. We can use the current method: pure AST node name comparison - the `log` method of `console` variable was called
> 
> 2. We can try to do some heuristics based on the `.d.ts` file path: for example, if it includes `node_modules` or something like that

Personally, I'm in favor of the second option, because it covers more cases and does so more reliably. On the other hand, pure AST checks have been around for a long time and no one has complained about them yet.

What do you think of that?

Reply:
Similarly to my opinion in https://github.com/typescript-eslint/typescript-eslint/pull/8952#discussion_r1667775812, I think that Option 2 here is just out of scope for _this_ issue, granted that existing checks are based on Option 1 -like behavior. So my preference would be to go with Option 1 and discuss Option 2 in a followup.